File: Tutorials/Codes/Modern_Control_Course/Chapter13/Lesson2/Chapter13_Lesson2.py
# Chapter13_Lesson2.py
# Observable states and observable subspace for a continuous-time LTI system.
# Dependencies: numpy only. Optional modern-control packages: scipy.signal and python-control
# can be used for state-space simulation, but the subspace computation below is from scratch.

import numpy as np


def observability_signature_matrix(A: np.ndarray, C: np.ndarray) -> np.ndarray:
    """Build O = [C; C A; ...; C A^(n-1)]."""
    A = np.asarray(A, dtype=float)
    C = np.asarray(C, dtype=float)
    n = A.shape[0]
    blocks = []
    Ak = np.eye(n)
    for _ in range(n):
        blocks.append(C @ Ak)
        Ak = Ak @ A
    return np.vstack(blocks)


def numerical_rank(M: np.ndarray, tol: float = 1e-10) -> int:
    """SVD-based numerical rank."""
    s = np.linalg.svd(M, compute_uv=False)
    return int(np.sum(s > tol))


def null_space(M: np.ndarray, tol: float = 1e-10) -> np.ndarray:
    """Return an orthonormal basis for the null space of M as columns."""
    _, s, vh = np.linalg.svd(M, full_matrices=True)
    rank = int(np.sum(s > tol))
    return vh[rank:].T


def observable_projection(A: np.ndarray, C: np.ndarray, tol: float = 1e-10):
    """Compute the observable and unobservable subspace projectors.

    The unobservable subspace is ker(O). The observable subspace, in the
    standard Euclidean inner product, is ker(O)^perp = range(O.T).
    """
    O = observability_signature_matrix(A, C)
    N = null_space(O, tol)
    # SVD row-space basis for range(O.T)
    u, s, vh = np.linalg.svd(O, full_matrices=True)
    rank = int(np.sum(s > tol))
    V_o = vh[:rank].T
    P_o = V_o @ V_o.T
    P_u = N @ N.T if N.size else np.zeros((A.shape[0], A.shape[0]))
    return O, V_o, N, P_o, P_u


def output_signature(A: np.ndarray, C: np.ndarray, x0: np.ndarray) -> np.ndarray:
    """Finite derivative signature [C x0, C A x0, ..., C A^(n-1) x0]."""
    O = observability_signature_matrix(A, C)
    return O @ np.asarray(x0, dtype=float)


if __name__ == "__main__":
    # Example: x3 is dynamically separated from the measured coordinates.
    A = np.array([[0.0, 1.0, 0.0],
                  [-2.0, -3.0, 0.0],
                  [0.0, 0.0, -4.0]])
    C = np.array([[1.0, 0.0, 0.0]])

    O, V_o, N_u, P_o, P_u = observable_projection(A, C)

    print("Observability signature matrix O:")
    print(O)
    print("rank(O) =", numerical_rank(O))
    print("Observable subspace basis columns (range of O.T):")
    print(V_o)
    print("Unobservable subspace basis columns (ker O):")
    print(N_u)

    x0 = np.array([2.0, -1.0, 5.0])
    x_obs = P_o @ x0
    x_unobs = P_u @ x0
    print("x0 =", x0)
    print("observable component =", x_obs)
    print("unobservable component =", x_unobs)
    print("finite output derivative signature of x0 =", output_signature(A, C, x0))
    print("signature of observable component =", output_signature(A, C, x_obs))
    print("signature of unobservable component =", output_signature(A, C, x_unobs))

File: Tutorials/Codes/Modern_Control_Course/Chapter13/Lesson2/Chapter13_Lesson2.cpp
// Chapter13_Lesson2.cpp
// Observable states and observable subspace for an LTI system.
// Dependency: Eigen (https://eigen.tuxfamily.org). Compile for example with:
// g++ Chapter13_Lesson2.cpp -I /path/to/eigen -O2 -std=c++17 -o Chapter13_Lesson2

#include <Eigen/Dense>
#include <iostream>
#include <vector>

using Eigen::MatrixXd;
using Eigen::VectorXd;

MatrixXd observabilitySignatureMatrix(const MatrixXd& A, const MatrixXd& C) {
    const int n = static_cast<int>(A.rows());
    const int p = static_cast<int>(C.rows());
    MatrixXd O(p * n, n);
    MatrixXd Ak = MatrixXd::Identity(n, n);
    for (int k = 0; k < n; ++k) {
        O.block(k * p, 0, p, n) = C * Ak;
        Ak = Ak * A;
    }
    return O;
}

int numericalRank(const MatrixXd& M, double tol = 1e-10) {
    Eigen::JacobiSVD<MatrixXd> svd(M);
    int r = 0;
    for (int i = 0; i < svd.singularValues().size(); ++i) {
        if (svd.singularValues()(i) > tol) ++r;
    }
    return r;
}

MatrixXd nullSpace(const MatrixXd& M, double tol = 1e-10) {
    Eigen::JacobiSVD<MatrixXd> svd(M, Eigen::ComputeFullV);
    const auto& s = svd.singularValues();
    int r = 0;
    for (int i = 0; i < s.size(); ++i) {
        if (s(i) > tol) ++r;
    }
    return svd.matrixV().rightCols(M.cols() - r);
}

MatrixXd rowSpaceBasisAsColumns(const MatrixXd& M, double tol = 1e-10) {
    Eigen::JacobiSVD<MatrixXd> svd(M, Eigen::ComputeFullV);
    const auto& s = svd.singularValues();
    int r = 0;
    for (int i = 0; i < s.size(); ++i) {
        if (s(i) > tol) ++r;
    }
    return svd.matrixV().leftCols(r);
}

int main() {
    MatrixXd A(3, 3);
    A << 0.0, 1.0, 0.0,
        -2.0, -3.0, 0.0,
         0.0, 0.0, -4.0;

    MatrixXd C(1, 3);
    C << 1.0, 0.0, 0.0;

    MatrixXd O = observabilitySignatureMatrix(A, C);
    MatrixXd Nu = nullSpace(O);
    MatrixXd Vo = rowSpaceBasisAsColumns(O);

    std::cout << "Observability signature matrix O:\n" << O << "\n\n";
    std::cout << "rank(O) = " << numericalRank(O) << "\n\n";
    std::cout << "Observable subspace basis columns (range of O^T):\n" << Vo << "\n\n";
    std::cout << "Unobservable subspace basis columns (ker O):\n" << Nu << "\n\n";

    VectorXd x0(3);
    x0 << 2.0, -1.0, 5.0;
    MatrixXd Po = Vo * Vo.transpose();
    MatrixXd Pu = Nu.cols() > 0 ? Nu * Nu.transpose() : MatrixXd::Zero(3, 3);

    std::cout << "x0:\n" << x0 << "\n\n";
    std::cout << "observable component:\n" << Po * x0 << "\n\n";
    std::cout << "unobservable component:\n" << Pu * x0 << "\n";
    return 0;
}
